### PR TITLE
Orchestration file generation and metadata

### PR DESCRIPTION
**Pull request by atengler**

### CODE REVIEW

File: salt/files/orchestrate.sls [hunk 1]
```diff
@@ -1,36 +1,116 @@
...
+#}
+
+{%- for state in sorted_priorities %}
+
+{%- set formula = state.0.split('.') %}
 
-{%- set sorted_formula_list = new_formula_dict|dictsort(false, 'value') %}
-	
-{%- for formula in sorted_formula_list %}
+{%- if salt['file.directory_exists']('/srv/salt/env/'+environment_name+'/'+formula.0+'/orchestrate') and formula|length > 1 and salt['file.file_exists']('/srv/salt/env/'+environment_name+'/'+formula.0+'/orchestrate/'+formula.1+'.sls') %}
```
Comment (fpytloun) [inline salt/files/orchestrate.sls]: I don't like hard-coded `/srv/salt/env` here.. isn't there some parameter in map.jinja or something for this location?

Comment (fpytloun) [reply salt/files/orchestrate.sls]: The first conditional on existance of orchestrate directory is not needed as later you are testing existence of file in that directory.

File: salt/files/orchestrate.sls [hunk 2]
```diff
@@ -1,36 +1,116 @@
...
 
-{{ salt['cmd.run']('cat /srv/salt/env/'+environment_name+'/'+formula.0+'/orchestrate.sls') }}
+{%- else %}
+
+{%- if args[ state.0 ] is defined %}
+
+{{ state.0 }}:
+  salt.state:
+    - tgt: '{{ state.0|replace(".", ":") }}'
+    - tgt_type: pillar
```
Comment (fpytloun) [inline salt/files/orchestrate.sls:41]: Isn't it better to use compound here instead?

Comment (fpytloun) [inline salt/files/orchestrate.sls]: + shouldn't this be executed only if debug is false (according to the code bellow)?

File: salt/files/orchestrate.sls [hunk 3]
```diff
@@ -1,36 +1,116 @@
...
+{{ state.0 }}.logok:
+  salt.function:
+    - tgt: 'I@salt:master or I@{{ state.0|replace(".", ":") }}'
+    - tgt_type: compound
+    - queue: True
+    - name: cmd.run
+    - arg:
+      - 'echo "$(date +"%d %h %Y %H:%M:%S") | state: {{ state.0}} - status: OK" >> /var/log/salt/orchestrate_runner'
+    - require:
+      - salt: {{ state.0 }}
```
Comment (fpytloun) [inline salt/files/orchestrate.sls:57]: Shouldn't be debug executed even if state fails?

File: salt/master/orchestrate.sls [hunk 4]
```diff
@@ -5,28 +5,67 @@
...
-{% _dummy = formula_dict.update{formula_name: grains_yaml.orchestrate }}
+
+{%- for state, priority in grains_yaml['orchestrate'].iteritems() %}
+
+{%- do priorities.update({ formula_name+'.'+state: grains_yaml['orchestrate'][state]['priority'] }) %}
+
+{%- set arguments = [] %}
+{%- for arg_name, arg_value in grains_yaml['orchestrate'][state].iteritems() %}
+
+{%- if 'priority' not in arg_name %}
```
Comment (fpytloun) [inline salt/master/orchestrate.sls]: Indenting would be really helpful in this code

File: salt/files/orchestrate.sls (hunk 3 above)
Comment (atengler) [reply salt/files/orchestrate.sls:57]: It is executed by logfail state under this state

File: salt/files/orchestrate.sls (hunk 2 above)
Comment (atengler) [reply salt/files/orchestrate.sls:41]: Compound allows you to match minion using multiple methods (grain, pillar, minion ID etc. all together), there is no other method I need to (or even can) use, most usable grains items (role for instance) are only generated after salt state ran on the minion, so pillar matching felt more suitable (its always available) and minion IDs can be different on various deployments

File: salt/files/orchestrate.sls (hunk 1 above)
Comment (atengler) [reply salt/files/orchestrate.sls]: I'll look at this one later today

File: salt/master/orchestrate.sls (hunk 4 above)
Comment (atengler) [reply salt/master/orchestrate.sls]: OK, it would make it more readable, I'll look at this later today

File: salt/files/orchestrate.sls (hunk 2 above)
Comment (fpytloun) [reply salt/files/orchestrate.sls:41]: And what about `and not` and other cases? I guess these are only available in compound mode.

Comment (atengler) [reply salt/files/orchestrate.sls:41]: Can you provide any specific use case?

**Comment by cznewt** [comment]
Nice work, little penguin

**Pull request merged**